cheroot/ssl/builtin.py

```diff
@@ -173,22 +239,164 @@ def get_environ(self, sock):
+        with suppress(AttributeError):
+            compression = sock.compression()
+            if compression is not None:
+                ssl_environ['SSL_COMPRESS_METHOD'] = compression
+
+        # Python 3.6+
+        with suppress(AttributeError):
+            ssl_environ['SSL_SESSION_ID'] = sock.session.id.hex()
+        with suppress(AttributeError):
+            target_cipher = cipher[:2]
+            for cip in sock.context.get_ciphers():
+                if target_cipher == (cip['name'], cip['protocol']):
+                    ssl_environ['SSL_CIPHER_ALGKEYSIZE'] = cip['alg_bits']
+                    break
+
+        # Python 3.7+ sni_callback
+        with suppress(AttributeError):
+            ssl_environ['SSL_TLS_SNI'] = sock.sni
+
         if self.context and self.context.verify_mode != ssl.CERT_NONE:
             client_cert = sock.getpeercert()
             if client_cert:
-                for cert_key, env_var in self.CERT_KEY_TO_ENV.items():
-                    ssl_environ.update(
-                        self.env_dn_dict(env_var, client_cert.get(cert_key)),
-                    )
+                ssl_environ['SSL_CLIENT_VERIFY'] = 'SUCCESS'
+                ssl_environ.update(
+                    self.env_cert_dict('SSL_CLIENT', client_cert),
+                )
+                ssl_environ['SSL_CLIENT_CERT'] = ssl.DER_cert_to_PEM_cert(
+                    sock.getpeercert(binary_form=True),
+                ).strip()
+
+        ssl_environ.update(self._server_env)
+
+        # not supplied by the Python standard library (as of 3.8)
+        # - SSL_SESSION_RESUMED
+        # - SSL_SECURE_RENEG
+        # - SSL_CLIENT_CERT_CHAIN_n
+        # - SRP_USER
+        # - SRP_USERINFO
 
         return ssl_environ
 
+    def loopback_for_cert(self, cert, private_key):
+        """Create a loopback connection to parse a cert with a private key."""
+        context = ssl.create_default_context(
+            cafile=self.certificate_chain,
+        )
+        context.load_cert_chain(cert, private_key)
+        context.check_hostname = False
+        context.verify_mode = ssl.CERT_NONE
+
+        # Python 3+ Unix, Python 3.5+ Windows
+        client, server = socket.socketpair()
+        try:
+            # ssl handshake is blocking and needs to happen on
+            # both ends at the same time so use a thread
+            thread = threading.Thread(
+                target=lambda: context.wrap_socket(
+                    server, do_handshake_on_connect=True,
+                    server_side=True,
+                ),
+            )
+            try:
+                thread.start()
+                return context.wrap_socket(
+                    client, do_handshake_on_connect=True,
+                    server_side=False,
+                ).getpeercert()
+            finally:
+                thread.join()
+        finally:
+            client.close()
+            server.close()
+
+    def parse_cert(self, cert, private_key):
```

Comment:
`self` is only used once. This method could be just a helper function.

Reply:
This would allow us to unit-test it w/o need to mock any side-effects.

Reply:
pulled off the class along with `loopback_for_cert`